### ISSUE TITLE
CS4014 not triggered for non-explicitly async methods

### ISSUE THREAD

**Issue by onyxmaster** [dotnet/roslyn#20782]
**Version Used**: 
VS 2017 15.3.0 Preview 3.0

**Steps to Reproduce**:
```c#
using System.Threading.Tasks;

static class Program
{
    static void Main()
    {
        F1();
        F2();
    }

    static Task F1() => Task.CompletedTask;

    static async Task F2() { await Task.CompletedTask; }
}
```

**Expected Behavior**:
Both lines with `F1()` and `F2()` should trigger CS4014.

**Actual Behavior**:
Only the `F2()` triggers CS4014.

**Other Info**:
This is mostly apparent when consuming an asynchronous interface, where one cannot mark the method as asynchronous explicitly.

**Comment by sharwell** [comment]
💭 I actually wasn't expecting to see CS4014 anywhere here, but if it doesn't get reported for `F2()` then I agree it makes sense to also report it for `F1()`.

**Comment by jaredpar** [comment]
This behavior is "By Design". The compiler will only issue CS4014 for explicitly marked async methods. Having it issue the warning for all methods that return `Task` is a significant compat break.

**Comment by m-akinc** [comment]
I can appreciate maintaining compatibility, but is there an alternative way to address this? Covering only calls to explicitly async methods seems like it leaves out a significant portion of the problem cases, and the motivation for that distinction isn't apparent to me. Could they be covered by a new, separate warning? Or maybe introduce a configuration option to modify the behavior of the warning (e.g. like what exists for the stylecop analyzers: https://github.com/DotNetAnalyzers/StyleCopAnalyzers/blob/master/documentation/Configuration.md)?

**Comment by sharwell** [comment]
@m-akinc I would suggest proposing a new analyzer in [DotNetAnalyzers/AsyncUsageAnalyzers](https://github.com/DotNetAnalyzers/AsyncUsageAnalyzers).

**Comment by roji** [comment]
@sharwell @jaredpar this just bit us in EF Core (https://github.com/dotnet/efcore/issues/26790) - due to a copy-paste error, a synchronous method ended up unintentionally invoking an async method instead of the proper sync one.

I can see [VSTHRD110](https://github.com/Microsoft/vs-threading/blob/main/doc/analyzers/VSTHRD110.md), though I'm not sure if that ever got in. Even if it did, doesn't this belong in the built-in diagnostics analyzers, especially now that we have warning waves which solve the backwards compat issue? This seems like quite a pit of failure.

**Comment by rdewaele** [comment]
We (CC @databrecht) also got bit by this during refactoring of controller methods, which made our controller tests randomly fail. Changing a regular method to an async method is rather error-prone when not able to rely on the analyzers to flag these kinds of problems. In some projects such as a project that relies on EF core I'd be happy to be able to flag an async method call in a non-async context as undesired. 

In a sense it also reminds me of nullable reference types, where a change in semantics is guarded behind a feature flag. I am not sure this exact approach would make sense here, but it is an example where a project has some control about language semantics.

**Comment by databrecht** [comment]
For others who might find this issue, the following analyzer successfully catches the problem in our case
https://www.nuget.org/packages/Lindhart.Analyser.MissingAwaitWarning/2.1.0-beta
It also triggers warnings for F1 in the above example.

**Comment by indy-singh** [comment]
We just got bitten by this, and it wasn't great to debug.

> Having it issue the warning for all methods that return Task is a significant compat break.

Is it at all possible to have a CS4014_V2 that is opt in and we accept the burden of compat break?

For now we are opting to use [LindhartAnalyserMissingAwaitWarningStrict](https://github.com/ykoksen/unused-task-warning/blob/master/Lindhart.Analyser.MissingAwaitWarning/Lindhart.Analyser.MissingAwaitWarning/Lindhart.Analyser.MissingAwaitWarning/LindhartAnalyserMissingAwaitWarningAnalyzer.cs). 

Cheers,
Indy

**Comment by LuminoDiode** [comment]
> it wasn't great to debug

Agreed. Even if we look at common EFCore call:
`var user = _context.Users.FirstOrDefaultAsync(`
I want to see warning here. No way someone really wants to get task here.

**Comment by CyrusNajmabadi** [comment]
> I want to see warning here.

You can add your own warnings by writing an analyzer.

**Comment by stijnherreman** [comment]
Trying to introduce async to a mostly sync codebase is often a terrible experience. Sometimes I even end up reverting to sync when the tree of affected methods gets larger than I'm comfortable with, because my only tool is 'Find all references' and I don't have unit tests or integration tests or compiler warnings. Am I missing something that trivialises this kind of refactor?

> You can add your own warnings by writing an analyzer.

Wearing my dotnet-dev-writing-LOB-code hat, that's kind of a big ask.
